ttl/ttlworker/task_manager.go

```diff
@@ -0,0 +1,483 @@
+			workers = append(workers, w.(scanWorker))
+		}
+	}
+	return workers
+}
+
+func (m *taskManager) rescheduleTasks(se session.Session, now time.Time) {
+	idleScanWorkers := m.idleScanWorkers()
+	if len(idleScanWorkers) == 0 {
+		return
+	}
+
+	for len(idleScanWorkers) > 0 {
+		tasks, err := m.peekWaitingScanTasks(se, len(idleScanWorkers), now)
+		if err != nil {
+			logutil.Logger(m.ctx).Warn("fail to peek scan task", zap.Error(err))
+			return
+		}
+
+		if len(tasks) == 0 {
+			break
+		}
+
+		for _, t := range tasks {
+			logger := logutil.Logger(m.ctx).With(zap.String("jobID", t.JobID), zap.Int64("scanID", t.ScanID))
+
+			task, err := m.lockScanTask(se, t, now)
+			if err != nil {
+				// If other nodes lock the task, it will return an error. It's expected
+				// so the log level is only `info`
+				logutil.Logger(m.ctx).Info("fail to lock scan task", zap.Error(err))
+				continue
+			}
+
+			idleWorker := idleScanWorkers[0]
+			idleScanWorkers = idleScanWorkers[1:]
+
+			err = idleWorker.Schedule(task)
+			if err != nil {
+				logger.Warn("fail to schedule task", zap.Error(err))
+				continue
+			}
+
+			logger.Info("scheduled ttl task")
+			m.runningTasks = append(m.runningTasks, &runningScanTask{
+				ttlScanTask: task,
+				result:      nil,
+			})
+		}
+	}
+}
+
+func (m *taskManager) peekWaitingScanTasks(se session.Session, limit int, now time.Time) ([]*cache.TTLTask, error) {
+	sql, args := cache.PeekWaitingTTLTask(limit, now.Add(-2*ttlTaskHeartBeatTickerInterval))
+	rows, err := se.ExecuteSQL(m.ctx, sql, args...)
+	if err != nil {
+		return nil, errors.Wrapf(err, "execute sql: %s", sql)
+	}
+
+	tasks := make([]*cache.TTLTask, 0, len(rows))
+	for _, r := range rows {
+		task, err := cache.RowToTTLTask(se, r)
+		if err != nil {
+			return nil, err
+		}
+		tasks = append(tasks, task)
+	}
+
+	return tasks, nil
+}
+
+func (m *taskManager) lockScanTask(se session.Session, task *cache.TTLTask, now time.Time) (*ttlScanTask, error) {
+	ctx := m.ctx
+
+	err := se.RunInTxn(ctx, func() error {
+		sql, args := cache.SelectFromTTLTaskWithID(task.JobID, task.ScanID)
+		rows, err := se.ExecuteSQL(ctx, sql, args...)
+		if err != nil {
+			return errors.Wrapf(err, "execute sql: %s", sql)
+		}
+		if len(rows) == 0 {
+			return errors.Errorf("didn't find task with jobID: %s, scanID: %d", task.JobID, task.ScanID)
+		}
+		task, err = cache.RowToTTLTask(se, rows[0])
+		if err != nil {
+			return err
+		}
+		if task.OwnerID != "" && !task.OwnerHBTime.Add(2*jobManagerLoopTickerInterval).Before(now) {
+			return errors.New("task is already scheduled")
+		}
+
+		sql, args = setTTLTaskOwnerSQL(task.JobID, task.ScanID, m.id, now)
+		_, err = se.ExecuteSQL(ctx, sql, args...)
+		if err != nil {
+			return errors.Wrapf(err, "execute sql: %s", sql)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	err = m.infoSchemaCache.Update(se)
+	if err != nil {
+		return nil, err
+	}
+	table, ok := m.infoSchemaCache.Tables[task.TableID]
```

Comment:
Is it better to check if table exists before transaction? Otherwise, the node will lock task successfully without executing it. The task will be retaken by other node when heartbeat timeout.

Reply:
Yes!

Reply:
Done